File: directory-watcher/src/main/scala/com/swoval/watcher/Flags.scala
package com.swoval.watcher

case class FileEvent(fileName: String, flags: Int, id: Long) extends Flags.Event {
  import Flags.Event._
  private def hasFlags(f: Int) = (flags & f) != 0
  def isModified = hasFlags(ItemInodeMetaMod | ItemModified)
  def isNewFile = !isModified && !isRemoved && !isTouched && hasFlags(ItemCreated)
  def isRemoved = hasFlags(ItemRemoved)
  def isTouched = hasFlags(ItemInodeMetaMod)
  override def toString() = s"FileEvent($fileName, $id, ${Flags.Event.flags(flags)})"
}

object Flags {
  case class Create(value: Int) {
    import Create._
    def setUseCFTypes = new Create(value | UseCFTypes)
    def setNoDefer = new Create(value | NoDefer)
    def setWatchRoot = new Create(value | WatchRoot)
    def setIgnoreSelf = new Create(value | IgnoreSelf)
    def setFileEvents = new Create(value | FileEvents)
    def setMarkSelf = new Create(value | MarkSelf)
    def useExtendedData = new Create(value | UseExtendedData)
  }

  object Create extends Create(0x00000002) {
    // https://developer.apple.com/documentation/coreservices/1455376-fseventstreamcreateflags?language=objc
    val None = 0
    val UseCFTypes = 0x00000001
    val NoDefer = 0x00000002
    val WatchRoot = 0x00000004
    val IgnoreSelf = 0x00000008
    val FileEvents = 0x00000010
    val MarkSelf = 0x00000020
    val UseExtendedData = 0x00000040
  }

  trait Event extends Any {
    //https://developer.apple.com/documentation/coreservices/1455361-fseventstreameventflags
    import Event._
    def flags: Int
    def isNone = flags == 0
    def mustScanSubDirs = (flags & MustScanSubDirs) != 0
    def userDropped = (flags & UserDropped) != 0
    def kernelDropped = (flags & KernelDropped) != 0
    def eventIdsWrapped = (flags & EventIdsWrapped) != 0
    def historyDone = (flags & HistoryDone) != 0
    def rootChanged = (flags & RootChanged) != 0
    def mount = (flags & Mount) != 0
    def unmount = (flags & Unmount) != 0
    def itemChangeOwner = (flags & ItemChangeOwner) != 0
    def itemCreated = (flags & ItemCreated) != 0
    def itemFinderInfoMod = (flags & ItemFinderInfoMod) != 0
    def itemInodeMetaMod = (flags & ItemInodeMetaMod) != 0
    def itemIsDir = (flags & ItemIsDir) != 0
    def itemIsFile = (flags & ItemIsFile) != 0
    def itemIsHardlink = (flags & ItemIsHardlink) != 0
    def itemIsLastHardlink = (flags & ItemIsLastHardlink) != 0
    def itemIsSymlink = (flags & ItemIsSymlink) != 0
    def itemModified = (flags & ItemModified) != 0
    def itemRemoved = (flags & ItemRemoved) != 0
    def itemRenamed = (flags & ItemRenamed) != 0
    def itemXattrMod = (flags & ItemXattrMod) != 0
    def ownEvent = (flags & OwnEvent) != 0
    def itemCloned = (flags & ItemCloned) != 0
  }
  object Event {
    implicit class Impl(val flags: Int) extends AnyVal with Event
    def apply(i: Int) = new Impl(i)
    def flags(flag: Event) = {
      import flag._
      val values = Seq(
        s"  mustScanSubDirs: ${mustScanSubDirs}",
        s"  userDropped: ${userDropped}",
        s"  kernelDropped: ${kernelDropped}",
        s"  eventIdsWrapped: ${eventIdsWrapped}",
        s"  historyDone: ${historyDone}",
        s"  rootChanged: ${rootChanged}",
        s"  mount: ${mount}",
        s"  unmount: ${unmount}",
        s"  itemChangeOwner: ${itemChangeOwner}",
        s"  itemCreated: ${itemCreated}",
        s"  itemFinderInfoMod: ${itemFinderInfoMod}",
        s"  itemInodeMetaMod: ${itemInodeMetaMod}",
        s"  itemIsDir: ${itemIsDir}",
        s"  itemIsFile: ${itemIsFile}",
        s"  itemIsHardlink: ${itemIsHardlink}",
        s"  itemIsLastHardlink: ${itemIsLastHardlink}",
        s"  itemIsSymlink: ${itemIsSymlink}",
        s"  itemModified: ${itemModified}",
        s"  itemRemoved: ${itemRemoved}",
        s"  itemRenamed: ${itemRenamed}",
        s"  itemXattrMod: ${itemXattrMod}",
        s"  ownEvent: ${ownEvent}",
        s"  itemCloned: ${itemCloned}",
      )
      s"EventStreamFlags(\n${values mkString "\n"}\n)"
    }
    val None = 0
    val MustScanSubDirs = 0x00000001
    val UserDropped = 0x00000002
    val KernelDropped = 0x00000004
    val EventIdsWrapped = 0x00000008
    val HistoryDone = 0x00000010
    val RootChanged = 0x00000020
    val Mount = 0x00000040
    val Unmount = 0x00000080
    val ItemChangeOwner = 0x00004000
    val ItemCreated = 0x00000100
    val ItemFinderInfoMod = 0x00002000
    val ItemInodeMetaMod = 0x00000400
    val ItemIsDir = 0x00020000
    val ItemIsFile = 0x00010000
    val ItemIsHardlink = 0x00100000
    val ItemIsLastHardlink = 0x00200000
    val ItemIsSymlink = 0x00040000
    val ItemModified = 0x00001000
    val ItemRemoved = 0x00000200
    val ItemRenamed = 0x00000800
    val ItemXattrMod = 0x00008000
    val OwnEvent = 0x00080000
    val ItemCloned = 0x00400000
  }
}

File: apple-file-system/src/main/java/com/swoval/watcher/Flags.java
package com.swoval.watcher;

public class Flags {

    public static class Create {
        public final int value;

        public Create(final int value) {
            this.value = value;
        }

        public int getValue() {
            return value;
        }

        public Create setUseCFTypes() {
            return new Create(value | UseCFTypes);
        }
        public Create setNoDefer() {
            return new Create(value | NoDefer);
        }
        public Create setWatchRoot() {
            return new Create(value | WatchRoot);
        }
        public Create setIgnoreSelf() {
            return new Create(value | IgnoreSelf);
        }
        public Create setFileEvents() {
            return new Create(value | FileEvents);
        }
        public Create setMarkSelf() {
            return new Create(value | MarkSelf);
        }
        public Create useExtendedData() {
            return new Create(value | UseExtendedData);
        }
        public static final int  None = 0;
        public static final int  UseCFTypes = 0x00000001;
        public static final int  NoDefer = 0x00000002;
        public static final int  WatchRoot = 0x00000004;
        public static final int  IgnoreSelf = 0x00000008;
        public static final int  FileEvents = 0x00000010;
        public static final int  MarkSelf = 0x00000020;
        public static final int  UseExtendedData = 0x00000040;
    }

    public interface Event {
        //https://developer.apple.com/documentation/coreservices/1455361-fseventstreameventflags
        int flags();
        default boolean isNone() {
            return flags() == 0;
        }
        default boolean mustScanSubDirs() {
            return (flags() & MustScanSubDirs) != 0;
        }
        default boolean userDropped() {
            return (flags() & UserDropped) != 0;
        }
        default boolean kernelDropped() {
            return (flags() & KernelDropped) != 0;
        }
        default boolean eventIdsWrapped() {
            return (flags() & EventIdsWrapped) != 0;
        }
        default boolean historyDone() {
            return (flags() & HistoryDone) != 0;
        }
        default boolean rootChanged() {
            return (flags() & RootChanged) != 0;
        }
        default boolean mount() {
            return (flags() & Mount) != 0;
        }
        default boolean unmount() {
            return (flags() & Unmount) != 0;
        }
        default boolean itemChangeOwner() {
            return (flags() & ItemChangeOwner) != 0;
        }
        default boolean itemCreated() {
            return (flags() & ItemCreated) != 0;
        }
        default boolean itemFinderInfoMod() {
            return (flags() & ItemFinderInfoMod) != 0;
        }
        default boolean itemInodeMetaMod() {
            return (flags() & ItemInodeMetaMod) != 0;
        }
        default boolean itemIsDir() {
            return (flags() & ItemIsDir) != 0;
        }
        default boolean itemIsFile() {
            return (flags() & ItemIsFile) != 0;
        }
        default boolean itemIsHardlink() {
            return (flags() & ItemIsHardlink) != 0;
        }
        default boolean itemIsLastHardlink() {
            return (flags() & ItemIsLastHardlink) != 0;
        }
        default boolean itemIsSymlink() {
            return (flags() & ItemIsSymlink) != 0;
        }
        default boolean itemModified() {
            return (flags() & ItemModified) != 0;
        }
        default boolean itemRemoved() {
            return (flags() & ItemRemoved) != 0;
        }
        default boolean itemRenamed() {
            return (flags() & ItemRenamed) != 0;
        }
        default boolean itemXattrMod() {
            return (flags() & ItemXattrMod) != 0;
        }
        default boolean ownEvent() {
            return (flags() & OwnEvent) != 0;
        }
        default boolean itemCloned() {
            return (flags() & ItemCloned) != 0;
        }
        int  None = 0;
        int  MustScanSubDirs = 0x00000001;
        int  UserDropped = 0x00000002;
        int  KernelDropped = 0x00000004;
        int  EventIdsWrapped = 0x00000008;
        int  HistoryDone = 0x00000010;
        int  RootChanged = 0x00000020;
        int  Mount = 0x00000040;
        int  Unmount = 0x00000080;
        int  ItemChangeOwner = 0x00004000;
        int  ItemCreated = 0x00000100;
        int  ItemFinderInfoMod = 0x00002000;
        int  ItemInodeMetaMod = 0x00000400;
        int  ItemIsDir = 0x00020000;
        int  ItemIsFile = 0x00010000;
        int  ItemIsHardlink = 0x00100000;
        int  ItemIsLastHardlink = 0x00200000;
        int  ItemIsSymlink = 0x00040000;
        int  ItemModified = 0x00001000;
        int  ItemRemoved = 0x00000200;
        int  ItemRenamed = 0x00000800;
        int  ItemXattrMod = 0x00008000;
        int  OwnEvent = 0x00080000;
        int  ItemCloned = 0x00400000;
        static String flags(Event flag) {
            StringBuilder builder = new StringBuilder();
            builder.append("\n  mustScanSubDirs: " + flag.mustScanSubDirs());
            builder.append("\n  userDropped: " + flag.userDropped());
            builder.append("\n  kernelDropped: " + flag.kernelDropped());
            builder.append("\n  eventIdsWrapped: " + flag.eventIdsWrapped());
            builder.append("\n  historyDone: " + flag.historyDone());
            builder.append("\n  rootChanged: " + flag.rootChanged());
            builder.append("\n  mount: " + flag.mount());
            builder.append("\n  unmount: " + flag.unmount());
            builder.append("\n  itemChangeOwner: " + flag.itemChangeOwner());
            builder.append("\n  itemCreated: " + flag.itemCreated());
            builder.append("\n  itemFinderInfoMod: " + flag.itemFinderInfoMod());
            builder.append("\n  itemInodeMetaMod: " + flag.itemInodeMetaMod());
            builder.append("\n  itemIsDir: " + flag.itemIsDir());
            builder.append("\n  itemIsFile: " + flag.itemIsFile());
            builder.append("\n  itemIsHardlink: " + flag.itemIsHardlink());
            builder.append("\n  itemIsLastHardlink: " + flag.itemIsLastHardlink());
            builder.append("\n  itemIsSymlink: " + flag.itemIsSymlink());
            builder.append("\n  itemModified: " + flag.itemModified());
            builder.append("\n  itemRemoved: " + flag.itemRemoved());
            builder.append("\n  itemRenamed: " + flag.itemRenamed());
            builder.append("\n  itemXattrMod: " + flag.itemXattrMod());
            builder.append("\n  ownEvent: " + flag.ownEvent());
            builder.append("\n  itemCloned: " + flag.itemCloned());
            return "EventStreamFlags(\n" + builder.toString() + "\n)";
        }
    }
}
